### PR TITLE
gltfio: Add internal MorphHelper, enable up to 255 targets.

### LINKED ISSUE #1487: support blend shapes more than 8
**Is your feature request related to a problem? Please describe.**

Our company HyprSense is making an avatar animation technology (like iPhone X Animoji). We were looking for rendering engines to showcase 3D characters, and they usually have 40~50 blend shapes rigged in.

With Unity / Unreal Engine / Apple SceneKit, we had no problem because their vertex shader supports that number of blend shapes for a skinned mesh renderer. We are having difficulties with Sceneform, where the number of blend shapes is limited to 8 (or 4 with blending normals).

We want you to ask if you have a plan to implement a feature to support 40+ blend shapes. 

**Describe the solution you'd like**
You might want to store the blend shape vertex and normal deformation information in a data texture, and load from the vertex shader to apply the deformation. There are similar threads discussing about this topic in the three.js repository, and they came up with the approach I have mentioned.

https://github.com/mrdoob/three.js/issues/15556
https://github.com/mrdoob/three.js/issues/14441

**OS and backend**
All OSes and backends

### PR DESCRIPTION
**Pull request by prideout**
horse2

This allows us to handle many more models, such as those exported by Modo, and the horse model in #1852.

We do this by picking the highest 4 weights. The indices for these weights are called the "primary indices", and we construct a new VertexBuffer for each unique set of primary indices.

In the near future this will be made more efficient by adding a new VertexBuffer API to the Filament core.

Fixes #1487.

### CODE REVIEW

**Review by romainguy — APPROVED** [review]
That's an awesome change!

File: libs/gltfio/src/FilamentAsset.cpp [hunk 1]
```diff
@@ -30,6 +30,8 @@ using namespace utils;
 namespace gltfio {
 
 FFilamentAsset::~FFilamentAsset() {
+    delete mAnimator;
```
Comment (pixelflinger) [inline libs/gltfio/src/FilamentAsset.cpp:33]: curious why this change? if order is important, maybe a comment?

File: libs/gltfio/src/MorphHelper.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        std::vector<Permutation> permutations;
+        const cgltf_node* sourceNode;
+    };
+
+    void addPermutation(Entity targetEntity, filament::math::ubyte4 primaryIndices);
+
+    filament::VertexBuffer* createVertexBuffer(const cgltf_primitive& prim, const UvMap& uvmap,
+            filament::math::ubyte4 primaryIndices);
+
+    bool mEnabled = true;
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.h:79]: move the bool at the end of the structure, to potentially avoid padding

File: libs/gltfio/src/MorphHelper.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * Internal class that partitions lists of morph weights and maintains a cache of VertexBuffer
+ * objects for each partition. This allows gltfio to support up to 255 morph targets.
+ *
+ * Each partition is associated with an unordered set of 4 (or fewer) morph target indices, which
+ * we call the "primary indices" for that time slice.
+ *
+ * Animator has ownership over a single instance of MorphHelper, thus it is 1:1 with FilamentAsset.
+ */
+class MorphHelper {
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.h:40]: add the `noexpect` decorator where needed (everywhere?)

File: libs/gltfio/src/MorphHelper.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * Picks the 4 most influential weights and applies them to the target entity.
+     */
+    void applyWeights(Entity targetEntity, float const* weights, size_t count);
+
+    /**
+     * Allows the caller to prime the MorphHelper cache without actually affecting any renderables.
+     */
+    void enableWrites(bool enable) { mEnabled = enable; }
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.h:54]: the description and name of the method seem very different? Also the name of the variable "mEnabled"

File: libs/gltfio/src/MorphHelper.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+private:
+    struct GltfPrim {
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.h:57]: `GltfPrimitive` (we try to not abbreviate names in our style guide)

File: libs/gltfio/src/MorphHelper.cpp [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+void MorphHelper::applyWeights(Entity entity, float const* weights, size_t count) {
+    auto renderableManager = &mAsset->mEngine->getRenderableManager();
+    auto renderable = renderableManager->getInstance(entity);
+
+    // If there are 4 or fewer targets, we can simply re-use the original VertexBuffer.
+    if (count <= 4) {
+        if (mEnabled) {
+            float4 vec(0, 0, 0, 0);
+            for (size_t i = 0; i < count; i++) vec[i] = weights[i];
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.cpp:75]: use { }

File: libs/gltfio/src/MorphHelper.cpp [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+void MorphHelper::applyWeights(Entity entity, float const* weights, size_t count) {
+    auto renderableManager = &mAsset->mEngine->getRenderableManager();
+    auto renderable = renderableManager->getInstance(entity);
+
+    // If there are 4 or fewer targets, we can simply re-use the original VertexBuffer.
+    if (count <= 4) {
+        if (mEnabled) {
+            float4 vec(0, 0, 0, 0);
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.cpp:74]: `float4 vec{}` -- slightly less verbose, not a biggie.

File: libs/gltfio/src/MorphHelper.cpp [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    for (size_t primIndex = 0; primIndex < permutation->primitives.size(); ++primIndex) {
+        const GltfPrim& prim = permutation->primitives[primIndex];
+        renderableManager->setGeometryAt(renderable, primIndex, prim.type, prim.vertices,
+               prim.indices, 0, prim.indices->getIndexCount());
+    }
+
+    // Finally, set the 4-tuple uniform for the weight values by derefing the primary indices.
+    // Note that we first create a "safe set" by replacing the unused sentinel with zero.
+    ubyte4 safe = permutation->primaryIndices;
+    for (int i = 0; i < 4; i++) if (safe[i] == kUnused) safe[i] = 0;
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.cpp:137]: omg! use `{}` here :-)

File: libs/gltfio/src/MorphHelper.cpp [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+
+    for (cgltf_size pi = 0; pi < prim_count; ++pi) {
+        const cgltf_primitive& prim = prims[pi];
+        RenderableManager::PrimitiveType prim_type;
+        getPrimitiveType(prim.type, &prim_type);
+        const auto& gltfioPrim = mAsset->mMeshCache.at(mesh)[pi];
+        VertexBuffer* const vertices = result.owner ?
+                createVertexBuffer(prim, gltfioPrim.uvmap, primaryIndices) :
+                gltfioPrim.vertices;
+        result.primitives.push_back({
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.cpp:165]: could it be `emplace_back` here?

File: libs/gltfio/src/MorphHelper.cpp [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+    }();
+
+    // This creates a copy because we don't know when the user will free the cgltf source data.
+    // For non-morphed vertex buffers, we use a sharing mechanism to prevent copies, but here
+    // we just want to keep it as simple as possible.
+    auto createBufferDescriptor = [](const cgltf_accessor* accessor) {
+        auto bufferData = (const uint8_t*) accessor->buffer_view->buffer->data;
+        const uint8_t* data = computeBindingOffset(accessor) + bufferData;
+        const uint32_t size = computeBindingSize(accessor);
+        uint8_t* clone = (uint8_t*) malloc(size);
```
Comment (pixelflinger) [inline libs/gltfio/src/MorphHelper.cpp:205]: this makes me very nervous, the `malloc` happens at a different place frmo the `free`. Could this be refactored to ahve the malloc/free in the same object, or, use a `unique_ptr`?

**Review by pixelflinger — CHANGES_REQUESTED** [review]

File: libs/gltfio/src/FilamentAsset.cpp (hunk 1 above)
Comment (prideout) [reply libs/gltfio/src/FilamentAsset.cpp:33]: Accidental movement, reverted.

File: libs/gltfio/src/MorphHelper.h (hunk 2 above)
Comment (prideout) [reply libs/gltfio/src/MorphHelper.h]: Done.

File: libs/gltfio/src/MorphHelper.h (hunk 4 above)
Comment (prideout) [reply libs/gltfio/src/MorphHelper.h]: Reworded the comment, also renamed the method to `disableWrites`.

File: libs/gltfio/src/MorphHelper.h (hunk 5 above)
Comment (prideout) [reply libs/gltfio/src/MorphHelper.h]: Done.

File: libs/gltfio/src/MorphHelper.cpp (hunk 7 above)
Comment (prideout) [reply libs/gltfio/src/MorphHelper.cpp]: Don.e

File: libs/gltfio/src/MorphHelper.cpp (hunk 9 above)
Comment (prideout) [reply libs/gltfio/src/MorphHelper.cpp]: Done.

File: libs/gltfio/src/MorphHelper.cpp (hunk 8 above)
Comment (prideout) [reply libs/gltfio/src/MorphHelper.cpp]: Done :)

File: libs/gltfio/src/MorphHelper.cpp (hunk 10 above)
Comment (prideout) [reply libs/gltfio/src/MorphHelper.cpp:210]: The `free` is actually just two lines after the malloc but you may've missed it because it was wrapped.  I fixed this to be more obvious.  (emscripten does not let me pass `free` directly into the callback)

Comment (pixelflinger) [reply libs/gltfio/src/MorphHelper.cpp:210]: ah yes... duh.

**Review by pixelflinger — APPROVED** [review]
